### ISSUE TITLE
UniquePtr and libMesh namespace

### ISSUE THREAD

**Issue by francesco-ballarin** [libMesh/libmesh#525]
Dear libmesh developers,
I saw the recent merge of the UniquePtr branch (PR #357) and now I would like to update my code with the new UniquePtr instead of AutoPtr.
Unfortunately, in my code I always use libMesh::AutoPtr, therefore I think a simple replace AutoPtr -> UniquePtr would not work, because of the current definition of UniquePtr as a macro.
For instance, in the disable-unique-ptr case, I guess libMesh::UniquePtr would be expanded to libMesh::libMesh::AutoPtr.

Even though I can of course replace in my code all libMesh::UniquePtr with UniquePtr, this would kind of defy the purpose of having the libMesh:: qualification on all the other libMesh-related objects of my code.

I created a fork with a patch, starting from the comments in the commits, which may provide a possible fix.
You can find it at https://github.com/francesco-ballarin/libmesh/commits/unique_ptr_namespace
I provide here a small summary of the patch, if you are willing to consider it at least for a discussion.

The patched auto_ptr.h is basically:

```
#if have c++11
use template aliases to define UniquePtr in the libMesh namespace
#elif have c++03
just copy the boost::unique_ptr implementation with the libMesh::UniquePtr name
#else // for instance, if disable-unique-ptr is on
just copy the old AutoPtr implementation with the new UniquePtr name
and provide a
#define AutoPtr UniquePtr
for backward compatibility
#endif
```

If this is helpful, I can (try to) create a P.R. However:
0. please check if this is the backward compatibility that you have in mind (there is no AutoPtr class in the enable-unique-ptr case)
1. in the #if we should also test if template aliases are supported by the compiler. I do not know how to do that.
2. contrib/unique_ptr/unique_ptr.hpp has been modified replacing the boost namespace with the libMesh one. After this modification, I wonder if it would be better to just move the content of this file inside auto_ptr.h (or move the file unique_ptr.h in the libMesh includes, rather then in the contrib folder).

Best regards,
Francesco

**Comment by roystgnr** [comment]
John, why use a macro rather than a typedef?

**Comment by roystgnr** [comment]
Oh, because template typedefs didn't make it into the language until C++11, of course...

**Comment by jwpeterson** [comment]
> On Apr 3, 2015, at 7:43 AM, roystgnr <notifications@github.com> wrote:
> 
> Oh, because template typedefs didn't make it into the language until C++11, of course...
> 

Right, I think I commented on this in a commit message... C++11 using declarations provide the ideal solution here, but we can't depend on them.  Macros were the only thing I could come up with that would allow existing code to compile but they don't work with libMesh:: namespaced code, unfortunately.

**Comment by francesco-ballarin** [comment]
I am sorry, I do not understand. Why can't we depend on them, if c++11 is enabled and (see also 1.) using declarations are supported by the compiler?

**Comment by permcody** [comment]
Because those typedefs scattered throughout the code won't work if you are
using a c++03 compiler. For now, libMesh is c++03 compliant so you can't
take advantage of any tricks that aren't backwards compatible.

**Comment by francesco-ballarin** [comment]
again, I do not understand. Of course we would like to mantain c++03 backward compatibility, and this is indeed taken care of by the #elif, I think.

**Comment by permcody** [comment]
The problem isn't on the definition side, it's on the expansion side in the
code itself.  You're welcome to try something out though. If you put a PR
up on Github, it will be tested by moosebuild.org which ensures C++03
compatibility.

Cody

**Comment by jwpeterson** [comment]
```
#if have c++11
use template aliases to define UniquePtr in the libMesh namespace
#elif have c++03
```

I should also point out `--{enable,disable}-cxx11` is mutually exclusive to `--{enable,disable}-unique-ptr, which may be another point of confusion here.  So, the logic would be something like

```
#if HAVE_CXX11_ALIAS_DECLARATIONS
#  ifdef LIBMESH_ENABLE_UNIQUE_PTR
#    ifdef LIBMESH_HAVE_CXX11_UNIQUE_PTR
#      include <memory>
       // C++11 using declaration to set UniquePtr == std::unique_ptr
#  elif LIBMESH_HAVE_HINNANT_UNIQUE_PTR
#    include "libmesh/unique_ptr.hpp"
     // C++11 using declaration to set UniquePtr == boost::unique_ptr
#  else
     // C++11 using declaration to set UniquePtr == libMesh::AutoPtr
#  endif
#else
  // C++11 using declaration to set UniquePtr == libMesh::AutoPtr
#endif


#else // don't have HAVE_CXX11_ALIAS_DECLARATIONS
// repeat the same set of #ifdefs as above, but using #define instead??? 
#endif
```

Something like this may actually work, but I haven't tried it yet.  Need to at least add an autoconf test for C++11 alias declarations... and it means that properly-namespaced `libMesh::UniquePtr` app code is C++11-only, which might be just fine depending on the app.

**Comment by roystgnr** [comment]
If you have full C++11 then we can use template typedefs.  And if you don't have *anything* then we might as well use a macro to define UniquePtr==AutoPtr, and that should work regardless of user namespace choices.  If you don't have full C++11 compatibility then I'm fine saying "no unique_ptr for you"; it's probably safer not to trust a half-implementation anyway.  But I don't want to say "you can't even use boost::unique_ptr".  Maybe we add a shim class just for the boost::unique_ptr-but-no-template-typedefs case?

**Comment by jwpeterson** [comment]
> On Apr 5, 2015, at 12:11 PM, francesco-ballarin <notifications@github.com> wrote:
> 
> I agree with Roy. I'll put my pull request in a few minutes, if it is helpful. I am not sure if it will work with moosebuild, I may need your help to fix it. In fact, (see point 2 in my original post), I renamed boost::unique_ptr to libMesh::UniquePtr
> 

I really don't want to do this, and see no reason why we should...

**Comment by roystgnr** [comment]
Resolved by #537 (and resolved even better by subsequent commits)
